## **Pandas Essencial**
**Prof. Dr. Samuel Martins (@hisamuka @xavecoding)** <br/>
xavecoding: https://youtube.com/c/xavecoding <br/>

Neste tutorial, vamos aprender o essencial da biblioteca _Pandas_ para manipulação de dados.<br/><br/>

<a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License</a>.

### Pacotes usados neste Notebook

In [ ]:
# pacotes usados neste notebook
import pandas as pd
from google.colab import drive
drive.mount('/content/drive')

# Caminho padrao para a base de dados
path = '/content/drive/MyDrive/trilia-dados/'

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


<h1>Manipulação de Dados com Pandas</h1><hr/>

<h2>1. Manipulação Básica de Datasets</h2>
<hr/>

**Dataset**: Gas Prices in Brazil: https://www.kaggle.com/matheusfreitag/gas-prices-in-brazil <br/>

Este dataset contém os **registros dos preços médios semanais dos combustíveis do Brasil entre os anos de 2004 e 2019**. <br/>
Cada *amostra (registro/linha)* consiste em um registro de preço aferido para um dado tipo de combustível em uma dada localidade do Brasil. <br/>
Alguns dos principais *atributos* (colunas) do dataset são: 'ESTADO', 'PRODUTO', 'NÚMERO DE POSTOS PESQUISADOS', 'PREÇO MÉDIO REVENDA'.


\* O arquivo disponibilizado no Kaggle está no formato *tsv*. Embora o _pandas_ consiga abrí-lo normalmente, convertemos tal arquivo para o formato *CSV*, que é um dos formatos mais utilizados, e mudamos seu separado para ';' apenas para mostrar algumas opções da função de carregamento.

### 1.1. Importando o Dataset
Para carregar um dataset no formato csv, basta utilizar a função `read_csv` do pandas. Por padrão, ela considera _','_ como separador.

In [ ]:
# Carregando o dataset corretamente ==> neste caso, usa o separador ';'
pd.read_csv('/content/drive/MyDrive/Trilha de Dados/GasPricesinBrazil_2004-2019 (1).csv')

,Unnamed: 0;DATA INICIAL;DATA FINAL;REGIÃO;ESTADO;PRODUTO;NÚMERO DE POSTOS PESQUISADOS;UNIDADE DE MEDIDA;PREÇO MÉDIO REVENDA;DESVIO PADRÃO REVENDA;PREÇO MÍNIMO REVENDA;PREÇO MÁXIMO REVENDA;MARGEM MÉDIA REVENDA;COEF DE VARIAÇÃO REVENDA;PREÇO MÉDIO DISTRIBUIÇÃO;DESVIO PADRÃO DISTRIBUIÇÃO;PREÇO MÍNIMO DISTRIBUIÇÃO;PREÇO MÁXIMO DISTRIBUIÇÃO;COEF DE VARIAÇÃO DISTRIBUIÇÃO;MÊS;ANO
0,0;2004-05-09;2004-05-15;CENTRO OESTE;DISTRITO ...
1,1;2004-05-09;2004-05-15;CENTRO OESTE;GOIAS;ETA...
2,2;2004-05-09;2004-05-15;CENTRO OESTE;MATO GROS...
3,3;2004-05-09;2004-05-15;CENTRO OESTE;MATO GROS...
4,4;2004-05-09;2004-05-15;NORDESTE;ALAGOAS;ETANO...
...,...
106818,106818;2019-06-23;2019-06-29;NORDESTE;RIO GRAN...
106819,106819;2019-06-23;2019-06-29;SUL;RIO GRANDE DO...
106820,106820;2019-06-23;2019-06-29;SUL;SANTA CATARIN...
106821,106821;2019-06-23;2019-06-29;SUDESTE;SAO PAULO...


O dataset não foi carregado corretamente pois o separador utilizado seu arquivo CSV era ';' e não a ','. <br/>
Vamos então carregá-lo corretamente:

In [ ]:
data = pd.read_csv('/content/drive/MyDrive/Trilha de Dados/GasPricesinBrazil_2004-2019 (1).csv', sep=';')
data

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,0.097,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,0.070,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,0.078,...,1.400,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
106818,106818,2019-06-23,2019-06-29,NORDESTE,RIO GRANDE DO NORTE,GNV,7,R$/m3,3.574,0.065,...,3.690,0.818,0.018,2.756,0,2.756,2.756,0,6,2019
106819,106819,2019-06-23,2019-06-29,SUL,RIO GRANDE DO SUL,GNV,23,R$/m3,3.401,0.129,...,3.789,0.95,0.038,2.451,0.402,1.9842,2.8661,0.164,6,2019
106820,106820,2019-06-23,2019-06-29,SUL,SANTA CATARINA,GNV,24,R$/m3,2.912,0.190,...,3.499,0.914,0.065,1.998,0,1.9981,1.9981,0,6,2019
106821,106821,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,GNV,52,R$/m3,3.020,0.229,...,3.490,0.646,0.076,2.374,0.165,2.0179,2.5093,0.07,6,2019


### 1.2. Exibindo as primeiras linhas do Dataset
A função `.head()` exibe as 5 primeiras linhas do dataset/tabela/Data Frame.

In [ ]:
# exibe as 10 primeiras linhas do dataframe
data.head(10)

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,0.097,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,0.070,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,0.078,...,1.400,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004
5,5,2004-05-09,2004-05-15,NORDESTE,BAHIA,ETANOL HIDRATADO,408,R$/l,1.383,0.132,...,2.050,0.426,0.095,0.957,0.128,0.5686,1.35,0.134,5,2004
6,6,2004-05-09,2004-05-15,NORDESTE,CEARA,ETANOL HIDRATADO,278,R$/l,1.453,0.218,...,1.950,0.353,0.150,1.1,0.068,0.7332,1.294,0.062,5,2004
7,7,2004-05-09,2004-05-15,NORDESTE,MARANHAO,ETANOL HIDRATADO,105,R$/l,1.631,0.158,...,2.000,0.515,0.097,1.116,0.091,0.9923,1.36183,0.082,5,2004
8,8,2004-05-09,2004-05-15,NORDESTE,PARAIBA,ETANOL HIDRATADO,125,R$/l,1.284,0.130,...,1.699,0.353,0.101,0.931,0.072,0.7577,1.2972,0.077,5,2004
9,9,2004-05-09,2004-05-15,NORDESTE,PERNAMBUCO,ETANOL HIDRATADO,423,R$/l,1.224,0.141,...,1.700,0.277,0.115,0.947,0.091,0.5643,1.3541,0.096,5,2004


### 1.3 Informações do Dataset e Elementos Chave

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 106823 entries, 0 to 106822
Data columns (total 21 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   Unnamed: 0                     106823 non-null  int64  
 1   DATA INICIAL                   106823 non-null  object 
 2   DATA FINAL                     106823 non-null  object 
 3   REGIÃO                         106823 non-null  object 
 4   ESTADO                         106823 non-null  object 
 5   PRODUTO                        106823 non-null  object 
 6   NÚMERO DE POSTOS PESQUISADOS   106823 non-null  int64  
 7   UNIDADE DE MEDIDA              106823 non-null  object 
 8   PREÇO MÉDIO REVENDA            106823 non-null  float64
 9   DESVIO PADRÃO REVENDA          106823 non-null  float64
 10  PREÇO MÍNIMO REVENDA           106823 non-null  float64
 11  PREÇO MÁXIMO REVENDA           106823 non-null  float64
 12  MARGEM MÉDIA REVENDA          

### 1.3.1 Informações gerais sobre o Dataset

A primeira coluna da tabela, chamada _'Unnamed: 0'_, parece não significar nada. Na verdade, ela parece ser os **índices** da tabela, que foram salvos como uma _coluna_.<br/>
Veremos jajá como **remover tal coluna**.

Outro ponto é que, aparentemente, nenhum atributo/coluna possui valores nulos (_null_), uma vez que o número de registros do dataframe e os números de valores _non-null_ é de **106823**. <br/>
Mas, veremos que não é bem assim para esse caso.

### 1.3.2 Data Frame
Todo dataset carregado (dados estruturados) é um `Data Frame`: 'Tabela' bi-dimensional, de tamanho mutável, com dados potencialmente heterogêneos. <br/>

In [ ]:
type(data)

pandas.core.frame.DataFrame

Podemos acessar as **dimensões do Data Frame** (número de linhas x número de colunas) utilizando o atributo `.shape` do Data Frame.

In [ ]:
data.shape

(106823, 21)

In [ ]:
print(f'O Dataframe possui {data.shape[0]} linhas/observações/registros e {data.shape[1]} colunas/atributos/variaveis.')

O Dataframe possui 106823 linhas/observações/registros e 21 colunas/atributos/variaveis.


#### **Criando um DataFrame**

Podemos criar um DataFrame a partir de um _dicionário_, onde cada **chave** possui uma **lista de elementos de igual tamanho**.<br/>
As **chaves** representam as **colunas** e **cada um dos valores de sua lista** representa o **valor da linha** correspondente para aquela coluna.

In [ ]:
alunos_df = pd.DataFrame({
    'nome': ['Luke Skywalker', 'Yoda', 'Palpatine'],
    'idade': [16, 1000, 70],
    'peso': [70, 15, 60],
    'eh jedi': [True, True, False]  # o nome das colunas podem ter espaços
})

In [ ]:
alunos_df

,nome,idade,peso,eh jedi
0,Luke Skywalker,16,70,True
1,Yoda,1000,15,True
2,Palpatine,70,60,False


In [ ]:
alunos_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   nome     3 non-null      object
 1   idade    3 non-null      int64 
 2   peso     3 non-null      int64 
 3   eh jedi  3 non-null      bool  
dtypes: bool(1), int64(2), object(1)
memory usage: 207.0+ bytes


#### **VEJA MAIS**
Criando um Data Frame a partir de um dicionário: https://www.geeksforgeeks.org/how-to-create-dataframe-from-dictionary-in-python-pandas/

#### **Renomeando as colunas de um DataFrame**
**===>** O método `DataFrame.columns` retorna uma ";lista" com os **nomes de todas as colunas** do data frame.

In [ ]:
alunos_df.columns

Index(['nome', 'idade', 'peso', 'eh jedi'], dtype='object')

In [ ]:
type(alunos_df.columns)

pandas.core.indexes.base.Index

<br/>

**===>** Para **renomear colunas** do data frame, utilize o método `DataFrame.rename`, que retorna uma _cópia_ do data frame com as as colunas renomeadas:

In [ ]:
alunos_df

,nome,idade,peso,eh jedi
0,Luke Skywalker,16,70,True
1,Yoda,1000,15,True
2,Palpatine,70,60,False


In [ ]:
alunos_df.rename(columns={
    'nome': 'Nome Completo',   # renomeia a coluna nome 'nome' para 'Nome Completo'
    'idade': 'Idade'}

SyntaxError: incomplete input (ipython-input-2363913446.py, line 3)

Para renomear o _próprio_ data frame em questão, utilize o parâmetro `inplace=True`:

In [ ]:
alunos_df.rename(columns={ 'nome': 'Nome Completo', 'idade': 'Idade'}, inplace=True)

<br/>

**===>** Uma outra forma de **renomear todas as listas** de um data frame é passar uma _lista_ com os novos nomes das colunas para o atributo `DataFrame.columns`:

In [ ]:
alunos_df.columns

In [ ]:
alunos_df.columns = ['NOME', 'IDADE', 'PESO','EH_JEDI']

In [ ]:
alunos_df

### 1.3.3 Series

Array uni-dimensional com os dados e rótulos de um eixo.

In [ ]:
data.head()

In [ ]:
# selecionando uma coluna inteira
data['ESTADO']

In [ ]:
# selecionando uma coluna inteira
# esta forma de acesso, só funciona para colunas com nomes sem espaços, acentos, etc (caracteres inválidos)
data.ESTADO

In [ ]:
type(data['ESTADO'])

In [ ]:
# selecionando a observação indexada no índice [1] do dataframe
data.iloc[1]

#### **Criando uma Series**

Podemos criar um DataFrame a partir de uma lista de elementos.

In [ ]:
pd.Series([5.5,6.0,9.5  ])

Podemos alterar o **nome dos índices** (veremos melhor jajá) e o **nome da Series** (o que ela representa):

In [ ]:
pd.Series([5.5,6.0,9.5  ], index=['prova 1', 'prova 2', 'projeto'], name ='Notas dos Luke SkyWalker')

#### **VEJA MAIS**
https://pandas.pydata.org/pandas-docs/stable/user_guide/dsintro.html

### 1.3.4 Atribuindo Dados

#### 1.3.4.1 Atribuindo constantes

In [ ]:
produto_view = data['PRODUTO']  # a series retornada refente à coluna, NÃO É UMA CÓPIA, mas sim, uma REFERÊNCIA/VIEW à coluna do dataframe
produto_view


In [ ]:
produto_copy_bkp = data['PRODUTO'].copy()  # retorna uma cópia da coluna 'PRODUTO'
produto_copy_bkp


In [ ]:

data['PRODUTO'] = 'Combustível'  # atribuindo o valor constante 'Combustível' para linha do dataframe na coluna 'PRODUTO'

In [ ]:
data.head()

In [ ]:
produto_view

In [ ]:
produto_copy_bkp

#### 1.3.4.2 Atribuindo listas ou series

In [ ]:
nrows, ncols = data.shape
nrows, ncols

In [ ]:
novos_produtos = [f'Produto {i}' for i in range(nrows)]
len(novos_produtos)

In [ ]:
# a quantidade de elementos da lista `novos_produtos` é igual ao número de linhas do dataframe
data['PRODUTO'] = novos_produtos

In [ ]:
data

In [ ]:
print(produto_view)
print('\n')

In [ ]:

# voltando para os produtos originais
data['PRODUTO'] = produto_copy_bkp  # produto_copy_bkp é uma Series

In [ ]:
data

#### 1.3.4.3 Criando novas colunas
Para **criar uma nova coluna** em um data frame, basta atribuirmos uma _lista/Series de valores_ a uma **nova 'chave'** do data frame. <br/>

**PS:** A _quantidade de valores_ da lista precisa ser **igual** ao _número de linhas/registros_ do data frame.

In [ ]:
# criando uma coluna a partir de um valor constante/default
# todas as linhas terão o mesmo valor para esta nova coluna
data['coluna sem nocao'] = 'DEFAULT'
data

In [ ]:
data['coluna a partir de lista'] = range(data.shape[0])
data

In [ ]:
# não funciona pq a quantidade de elementos da lista (a serem atribuídos a nova coluna) é diferente
# da quantidade de linhas do dataframe
data['nao funciona'] = [1, 2, 3]

<br/>

Outro exemplo:

In [ ]:
data.head()

In [ ]:
data['PREÇO MÉDIO REVENDA (dólares)'] = data['PREÇO MÉDIO REVENDA'] * 6.0
data

**PS:** Obviamente, a lógica correta em converter o preço dos combustíveis em reais para dólares não é considerar uma taxa de câmbio fixa, uma vez que cada preço foi aferido em um momento diferente.

### 1.3.4 Índices

Todo Data Frame possui **índices**, que não são considerado colunas da tabela. Tais índices são comumente **númericos**, de 0 a num_linhas-1, mas também podem ser **textuais (rótulos/labels)**.

In [ ]:
data

In [ ]:
data.index

Use `list(data.index)` ou `data.index.to_list()` para converter um RangeIndex para uma python list.

#### **Exemplo de Data Frame com Índices Textuais (labels)**

In [ ]:
pesquisa_de_satisfacao = pd.DataFrame({
    'bom': [50, 21, 100],
    'ruim': [131, 2, 30],
    'pessimo': [30, 20, 1]
}, index=['XboxOne', 'Playstation4', 'Switch'])

In [ ]:
pesquisa_de_satisfacao.head()

### 1.4 Selecionando uma ou mais amostras (Indexação)

In [ ]:
data.head()

#### **==>  Index-based selection (seleção baseada em Índices)**
Mostrando linhas específicas de um DataFrame:

`iloc`: seleciona elementos do Dataframe, baseado em seu **índice (número)** --> row-first, column-second

**Selecionando uma amostra/linha:**

In [ ]:
# selecionando a linha 1 ===> observação de índice [1] do dataframe
data.iloc[1]

In [ ]:
data.iloc[:6]

**Selecionando múltiplias amostras/linhas:**

In [ ]:
# selecionando as linhas de índice de 10 a 15 (incluso)
data.iloc[10:16]

In [ ]:
# selecionando as linhas/observações de índice 1, 5, 10, 15
data.iloc[[1, 5, 10, 15]]

In [ ]:
# selecionando as linhas/observações de índice 5, 1, 10, 15
data.iloc[[5, 1, 10, 15]]

In [ ]:
# retornar o valor da linha de índice 1, coluna 4 ('ESTADO')
data.iloc[1, 4]

In [ ]:
# e assim por diante!

#### **==>  Label-based selection (seleção baseadas em Rótulos)**

`loc`: seleciona elementos do Dataframe, baseado em seus **rótulos** --> row-first, column-second

In [ ]:
pesquisa_de_satisfacao

In [ ]:
# retorna o valor linha de índice 0 (implícito) e a coluna de índice 1 (implícito) ==> usando o iloc
pesquisa_de_satisfacao.iloc[0,1]

In [ ]:
# retorna a linha cujo o rótulo do índice é 'XboxOne'
pesquisa_de_satisfacao.loc['XboxOne']

In [ ]:
# NÃO FUNCIONA  ===> iloc tentando acessar índices com rótulos
pesquisa_de_satisfacao.iloc['XboxOne']

In [ ]:
# NÃO FUNCIONA  ===> iloc tentando acessar índices rotulados com números
pesquisa_de_satisfacao.loc['0']

In [ ]:
pesquisa_de_satisfacao


In [ ]:
# retorna o valor da linha 'Playstation4', coluna 'ruim'
pesquisa_de_satisfacao.loc['Playstation4','ruim']

In [ ]:
# selecionando as linhas  de índices  resultados 'XboxOne','Switch'
pesquisa_de_satisfacao.loc[['XboxOne','Switch']]

In [ ]:
# retorna todas as linhas e apenas as colunas com rótulos 'bom' e 'pessimo'
pesquisa_de_satisfacao[['bom', 'pessimo']]

In [ ]:
# retorna todas as linhas e apenas as colunas com rótulos 'bom' e 'pessimo'
pesquisa_de_satisfacao.loc[:,['bom', 'pessimo']]

### 1.5 Selecionando um ou mais atributos (colunas)

In [ ]:
data.head()

In [ ]:
# retornando a coluna/atributo 'ESTADO'
data['ESTADO']

In [ ]:
data.ESTADO

In [ ]:
data.loc[:, 'ESTADO']

In [ ]:
# a forma de acesso de colunas por .NOME_DA_COLUNA só funciona se
# NOME_DA_COLUNA não possuir caracteres   ínvalidos (espaços, acentos, cedilha, ...)
### NÃO FUNCIONA
data.DATA INICIAL

In [ ]:
# para colunas cujos rótulos  possuem caracteres ínvalidos,
# apenas para a seleção via string é válida
data['DATA INICIAL']

Como o rótulo da coluna 'DATA INICIAL' **possui espaço**, não é possível acessá-la como método: `data.DATA INICIAL`

In [ ]:
# selecionando mútiplas colunas: 'PRODUTO', 'ESTADO', 'REGIÃO'
data[['PRODUTO', 'ESTADO', 'REGIÃO']]

### 1.6 Removendo um Atributo (Coluna) do Data Frame

Como vimos anteriormente, o atributo 'Unnamed: 0' parece ser um **ruído** em nosso dataset. Desta maneira, vamos eliminá-lo,

In [ ]:
data.head()

In [ ]:
# delete/remove in-place (ou seja,no própra  dataframe) a coluna de rótulo 'VUnnamed: 0'
del data ['Unnamed: 0']

In [ ]:
data

In [ ]:
del data['coluna sem nocao']
del data['coluna a apartir de lista']
del data['PREÇO MÉDIO REVENDA (dólares)']

### 1.7 Salvando um Data Frame

Para salvarmos um Data Frame para um **arquivo CSV**, basta usarmos o método `.to_csv`. <br/>
Por padrão, esse método **salva os índices da tabela como uma coluna no CSV**.<br/>
Como no geral tais índices são números de 0 a n-1, não há necessidade para isso (veja que removemos anteriormente a coluna 'Unnamed: 0' que foi justamente esse caso).<br/>
Desta forma, utilize o parâmetro: `index=False`.

Por padrão, o método utilizará a ',' como separador das colunas. Caso queira alterá-lo, utilize o parâmetro `sep`.

In [ ]:
data.to_csv('/content/drive/MyDrive/Trilha de Dados/GasPricesinBrazil_2004-2019 (1).csv', index=False)

### 1.8 Seleção Condicional: Filtrando amostras

Durante nossas análise exploratórias, frequentemente filtraremos nossas amostras, a partir de certas **condições**, para fins de análise mais específica. <br/>
Existem algumas maneiras de fazermos tal filtragem. Antes disso, vamos carregar nosso dataset pré-processado que salvamos no item anterior.

In [ ]:
data = pd.read_csv('/content/drive/MyDrive/Trilha de Dados/GasPricesinBrazil_2004-2019 (1).csv')
data.head()

,Unnamed: 0.1,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
4,4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,...,1.400,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004


In [ ]:
# Mostra todos os estados cujos os preço dos combustíveis foram aferidos
# Mais tecnicamente , mostra os valores únicos presentes para atríbuto/coluna 'ESTADO'
data['ESTADO'].unique()

array(['DISTRITO FEDERAL', 'GOIAS', 'MATO GROSSO', 'MATO GROSSO DO SUL',
       'ALAGOAS', 'BAHIA', 'CEARA', 'MARANHAO', 'PARAIBA', 'PERNAMBUCO',
       'PIAUI', 'RIO GRANDE DO NORTE', 'SERGIPE', 'ACRE', 'AMAPA',
       'AMAZONAS', 'PARA', 'RONDONIA', 'RORAIMA', 'TOCANTINS',
       'ESPIRITO SANTO', 'MINAS GERAIS', 'RIO DE JANEIRO', 'SAO PAULO',
       'PARANA', 'RIO GRANDE DO SUL', 'SANTA CATARINA'], dtype=object)

#### **Selecionando apenas os preços dos Postos de São Paulo**

##### **==> Alternativa 1: Seleção Condicional (Comparações diretas)**

O código abaixo retorna uma ***Series ('array') de booleans***, com o número de linhas (amostras) do Data Frame, que informa os registros de preços dos postos do _estado de São Paulo_ (True).

In [ ]:
# faz uma comparação elemento a elemento da series, retornando uma Series de booleans
data['ESTADO'] == 'SÃO PAULO'

,ESTADO
0,False
1,False
2,False
3,False
4,False
...,...
106818,False
106819,False
106820,False
106821,False


In [ ]:
# salvando essa Series de booleans em uma varíavel
seleção = data['ESTADO'] == 'SÃO PAULO'
seleção

,ESTADO
0,False
1,False
2,False
3,False
4,False
...,...
106818,False
106819,False
106820,False
106821,False


In [ ]:
type(seleção)

pandas.core.series.Series

In [ ]:
seleção.shape

(106823,)

In [ ]:
data.shape

(106823, 22)

Para **filtrarmos** os registros de postos do estado de São Paulo:

In [ ]:
data[seleção]

,Unnamed: 0.1,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO


O resultado é um Data Frame com _apenas_ os registros desejados após a **filtragem**.<br/>
Podemos ainda utilizar o método `loc` para o mesmo fim:

In [ ]:
data.loc[seleção]

,Unnamed: 0.1,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO


##### **==> Alternativa 2: Utilizando o método `query`**

`query` filtra linhas de um DataFrame baseado em uma **query (pergunta)**.

In [ ]:
data.head()

,Unnamed: 0.1,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
4,4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,...,1.400,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004


In [ ]:
data.query('ESTADO == "SAO PAULO"')

,Unnamed: 0.1,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
23,23,23,2004-05-09,2004-05-15,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4047,R$/l,0.768,...,1.150,0.198,0.111,0.57,0.056,0.4271,1.4123,0.098,5,2004
50,50,50,2004-05-16,2004-05-22,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4066,R$/l,0.766,...,1.150,0.19,0.108,0.576,0.05,0.43204,0.9724,0.087,5,2004
77,77,77,2004-05-23,2004-05-29,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4077,R$/l,0.823,...,1.199,0.209,0.086,0.614,0.063,0.34951,1.412,0.103,5,2004
104,104,104,2004-05-30,2004-06-05,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4073,R$/l,0.887,...,1.359,0.216,0.085,0.671,0.075,0.3398,1.0669,0.112,5,2004
131,131,131,2004-06-06,2004-06-12,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4082,R$/l,0.894,...,1.199,0.205,0.079,0.689,0.072,0.47797,1.0669,0.104,6,2004
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
106724,106724,106724,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,ÓLEO DIESEL,863,R$/l,3.448,...,4.220,0.401,0.045,3.047,0.097,2.8599,3.3481,0.032,6,2019
106751,106751,106751,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,ÓLEO DIESEL S10,1277,R$/l,3.574,...,4.320,0.451,0.047,3.123,0.111,2.885,3.491,0.036,6,2019
106778,106778,106778,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,GASOLINA COMUM,1644,R$/l,4.142,...,4.999,0.429,0.044,3.713,0.097,3.52,4.0892,0.026,6,2019
106805,106805,106805,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,GLP,1017,R$/13Kg,66.893,...,89.900,17.679,0.091,49.214,5.071,40.19,72,0.103,6,2019


<br/>

Uma boa prática é **salvar o Data Frame filtrado em uma nova variável**. Isso simplifica a complexidade do código para futuras análise feita para os postos de São Paulo.

In [ ]:
postos_sp = data.query('ESTADO == "SAO PAULO"')
postos_sp

,Unnamed: 0.1,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
23,23,23,2004-05-09,2004-05-15,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4047,R$/l,0.768,...,1.150,0.198,0.111,0.57,0.056,0.4271,1.4123,0.098,5,2004
50,50,50,2004-05-16,2004-05-22,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4066,R$/l,0.766,...,1.150,0.19,0.108,0.576,0.05,0.43204,0.9724,0.087,5,2004
77,77,77,2004-05-23,2004-05-29,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4077,R$/l,0.823,...,1.199,0.209,0.086,0.614,0.063,0.34951,1.412,0.103,5,2004
104,104,104,2004-05-30,2004-06-05,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4073,R$/l,0.887,...,1.359,0.216,0.085,0.671,0.075,0.3398,1.0669,0.112,5,2004
131,131,131,2004-06-06,2004-06-12,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4082,R$/l,0.894,...,1.199,0.205,0.079,0.689,0.072,0.47797,1.0669,0.104,6,2004
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
106724,106724,106724,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,ÓLEO DIESEL,863,R$/l,3.448,...,4.220,0.401,0.045,3.047,0.097,2.8599,3.3481,0.032,6,2019
106751,106751,106751,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,ÓLEO DIESEL S10,1277,R$/l,3.574,...,4.320,0.451,0.047,3.123,0.111,2.885,3.491,0.036,6,2019
106778,106778,106778,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,GASOLINA COMUM,1644,R$/l,4.142,...,4.999,0.429,0.044,3.713,0.097,3.52,4.0892,0.026,6,2019
106805,106805,106805,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,GLP,1017,R$/13Kg,66.893,...,89.900,17.679,0.091,49.214,5.071,40.19,72,0.103,6,2019


In [ ]:
type(postos_sp)

pandas.core.frame.DataFrame

In [ ]:
postos_sp.shape

(4263, 22)

In [ ]:
postos_sp.head()

,Unnamed: 0.1,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
23,23,23,2004-05-09,2004-05-15,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4047,R$/l,0.768,...,1.150,0.198,0.111,0.57,0.056,0.4271,1.4123,0.098,5,2004
50,50,50,2004-05-16,2004-05-22,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4066,R$/l,0.766,...,1.150,0.19,0.108,0.576,0.05,0.43204,0.9724,0.087,5,2004
77,77,77,2004-05-23,2004-05-29,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4077,R$/l,0.823,...,1.199,0.209,0.086,0.614,0.063,0.34951,1.412,0.103,5,2004
104,104,104,2004-05-30,2004-06-05,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4073,R$/l,0.887,...,1.359,0.216,0.085,0.671,0.075,0.3398,1.0669,0.112,5,2004
131,131,131,2004-06-06,2004-06-12,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4082,R$/l,0.894,...,1.199,0.205,0.079,0.689,0.072,0.47797,1.0669,0.104,6,2004


#### **Selecionando registros de postos do Rio de Janeiro com Preços acima de 2 reais**

In [ ]:
data.head()

,Unnamed: 0.1,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
4,4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,...,1.400,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004


In [ ]:
data['ESTADO'] == 'RIO DE JANEIRO'

,ESTADO
0,False
1,False
2,False
3,False
4,False
...,...
106818,False
106819,False
106820,False
106821,False


In [ ]:
dataPREÇO MÉDIO REVENDA

,PREÇO MÉDIO REVENDA
0,False
1,False
2,False
3,False
4,False
...,...
106818,True
106819,True
106820,True
106821,True


In [ ]:
seleção = (data['ESTADO'] == 'RIO DE JANEIRO ') & (data['PREÇO MÉDIO REVENDA']> 2.0)
seleção

,0
0,False
1,False
2,False
3,False
4,False
...,...
106818,False
106819,False
106820,False
106821,False


Note que o resultado da seleção continua sendo uma _Series de booleans_ com o _mesmo número de linhas/amostras do DataFrame_, de modo que cada linha possuirá um valor booleano indicando se o posto é do Rio de Janeiro e o preço aferido do combustível é maior que 2 reais (True) ou não.

O símbolo **&** significa **AND** na comparação. Essa nomenclatura do python/pandas é diferente das nomenclaturas tradicionais (&&). <br/>
Similarmente:
- **|** representa o **OR** (não é ||)
- **~** representa o **NOT** (não é !)

In [ ]:
data[seleção]

,Unnamed: 0.1,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO


In [ ]:
postos_sp.reset_index()

,index,Unnamed: 0.1,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,23,23,23,2004-05-09,2004-05-15,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4047,R$/l,...,1.150,0.198,0.111,0.57,0.056,0.4271,1.4123,0.098,5,2004
1,50,50,50,2004-05-16,2004-05-22,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4066,R$/l,...,1.150,0.19,0.108,0.576,0.05,0.43204,0.9724,0.087,5,2004
2,77,77,77,2004-05-23,2004-05-29,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4077,R$/l,...,1.199,0.209,0.086,0.614,0.063,0.34951,1.412,0.103,5,2004
3,104,104,104,2004-05-30,2004-06-05,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4073,R$/l,...,1.359,0.216,0.085,0.671,0.075,0.3398,1.0669,0.112,5,2004
4,131,131,131,2004-06-06,2004-06-12,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4082,R$/l,...,1.199,0.205,0.079,0.689,0.072,0.47797,1.0669,0.104,6,2004
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4258,106724,106724,106724,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,ÓLEO DIESEL,863,R$/l,...,4.220,0.401,0.045,3.047,0.097,2.8599,3.3481,0.032,6,2019
4259,106751,106751,106751,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,ÓLEO DIESEL S10,1277,R$/l,...,4.320,0.451,0.047,3.123,0.111,2.885,3.491,0.036,6,2019
4260,106778,106778,106778,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,GASOLINA COMUM,1644,R$/l,...,4.999,0.429,0.044,3.713,0.097,3.52,4.0892,0.026,6,2019
4261,106805,106805,106805,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,GLP,1017,R$/13Kg,...,89.900,17.679,0.091,49.214,5.071,40.19,72,0.103,6,2019


In [ ]:
postos_sp.reset_index(drop=True)

,Unnamed: 0.1,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,23,23,2004-05-09,2004-05-15,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4047,R$/l,0.768,...,1.150,0.198,0.111,0.57,0.056,0.4271,1.4123,0.098,5,2004
1,50,50,2004-05-16,2004-05-22,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4066,R$/l,0.766,...,1.150,0.19,0.108,0.576,0.05,0.43204,0.9724,0.087,5,2004
2,77,77,2004-05-23,2004-05-29,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4077,R$/l,0.823,...,1.199,0.209,0.086,0.614,0.063,0.34951,1.412,0.103,5,2004
3,104,104,2004-05-30,2004-06-05,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4073,R$/l,0.887,...,1.359,0.216,0.085,0.671,0.075,0.3398,1.0669,0.112,5,2004
4,131,131,2004-06-06,2004-06-12,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4082,R$/l,0.894,...,1.199,0.205,0.079,0.689,0.072,0.47797,1.0669,0.104,6,2004
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4258,106724,106724,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,ÓLEO DIESEL,863,R$/l,3.448,...,4.220,0.401,0.045,3.047,0.097,2.8599,3.3481,0.032,6,2019
4259,106751,106751,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,ÓLEO DIESEL S10,1277,R$/l,3.574,...,4.320,0.451,0.047,3.123,0.111,2.885,3.491,0.036,6,2019
4260,106778,106778,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,GASOLINA COMUM,1644,R$/l,4.142,...,4.999,0.429,0.044,3.713,0.097,3.52,4.0892,0.026,6,2019
4261,106805,106805,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,GLP,1017,R$/13Kg,66.893,...,89.900,17.679,0.091,49.214,5.071,40.19,72,0.103,6,2019


In [ ]:
# ainda não foi alterado, precisamos usar o atributo/parâmetro inplace=True
postos_sp

,Unnamed: 0.1,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
23,23,23,2004-05-09,2004-05-15,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4047,R$/l,0.768,...,1.150,0.198,0.111,0.57,0.056,0.4271,1.4123,0.098,5,2004
50,50,50,2004-05-16,2004-05-22,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4066,R$/l,0.766,...,1.150,0.19,0.108,0.576,0.05,0.43204,0.9724,0.087,5,2004
77,77,77,2004-05-23,2004-05-29,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4077,R$/l,0.823,...,1.199,0.209,0.086,0.614,0.063,0.34951,1.412,0.103,5,2004
104,104,104,2004-05-30,2004-06-05,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4073,R$/l,0.887,...,1.359,0.216,0.085,0.671,0.075,0.3398,1.0669,0.112,5,2004
131,131,131,2004-06-06,2004-06-12,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4082,R$/l,0.894,...,1.199,0.205,0.079,0.689,0.072,0.47797,1.0669,0.104,6,2004
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
106724,106724,106724,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,ÓLEO DIESEL,863,R$/l,3.448,...,4.220,0.401,0.045,3.047,0.097,2.8599,3.3481,0.032,6,2019
106751,106751,106751,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,ÓLEO DIESEL S10,1277,R$/l,3.574,...,4.320,0.451,0.047,3.123,0.111,2.885,3.491,0.036,6,2019
106778,106778,106778,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,GASOLINA COMUM,1644,R$/l,4.142,...,4.999,0.429,0.044,3.713,0.097,3.52,4.0892,0.026,6,2019
106805,106805,106805,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,GLP,1017,R$/13Kg,66.893,...,89.900,17.679,0.091,49.214,5.071,40.19,72,0.103,6,2019


In [ ]:
postos_sp.reset_index(drop=True, inplace=True)
postos_sp

,Unnamed: 0.1,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,23,23,2004-05-09,2004-05-15,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4047,R$/l,0.768,...,1.150,0.198,0.111,0.57,0.056,0.4271,1.4123,0.098,5,2004
1,50,50,2004-05-16,2004-05-22,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4066,R$/l,0.766,...,1.150,0.19,0.108,0.576,0.05,0.43204,0.9724,0.087,5,2004
2,77,77,2004-05-23,2004-05-29,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4077,R$/l,0.823,...,1.199,0.209,0.086,0.614,0.063,0.34951,1.412,0.103,5,2004
3,104,104,2004-05-30,2004-06-05,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4073,R$/l,0.887,...,1.359,0.216,0.085,0.671,0.075,0.3398,1.0669,0.112,5,2004
4,131,131,2004-06-06,2004-06-12,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4082,R$/l,0.894,...,1.199,0.205,0.079,0.689,0.072,0.47797,1.0669,0.104,6,2004
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4258,106724,106724,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,ÓLEO DIESEL,863,R$/l,3.448,...,4.220,0.401,0.045,3.047,0.097,2.8599,3.3481,0.032,6,2019
4259,106751,106751,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,ÓLEO DIESEL S10,1277,R$/l,3.574,...,4.320,0.451,0.047,3.123,0.111,2.885,3.491,0.036,6,2019
4260,106778,106778,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,GASOLINA COMUM,1644,R$/l,4.142,...,4.999,0.429,0.044,3.713,0.097,3.52,4.0892,0.026,6,2019
4261,106805,106805,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,GLP,1017,R$/13Kg,66.893,...,89.900,17.679,0.091,49.214,5.071,40.19,72,0.103,6,2019


In [ ]:
# ALTERNATIVA MUITO COMUM
postos_sp = data.query('ESTADO == "SAO PAULO"').reset_index(drop=True)
postos_sp

,Unnamed: 0.1,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,23,23,2004-05-09,2004-05-15,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4047,R$/l,0.768,...,1.150,0.198,0.111,0.57,0.056,0.4271,1.4123,0.098,5,2004
1,50,50,2004-05-16,2004-05-22,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4066,R$/l,0.766,...,1.150,0.19,0.108,0.576,0.05,0.43204,0.9724,0.087,5,2004
2,77,77,2004-05-23,2004-05-29,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4077,R$/l,0.823,...,1.199,0.209,0.086,0.614,0.063,0.34951,1.412,0.103,5,2004
3,104,104,2004-05-30,2004-06-05,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4073,R$/l,0.887,...,1.359,0.216,0.085,0.671,0.075,0.3398,1.0669,0.112,5,2004
4,131,131,2004-06-06,2004-06-12,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4082,R$/l,0.894,...,1.199,0.205,0.079,0.689,0.072,0.47797,1.0669,0.104,6,2004
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4258,106724,106724,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,ÓLEO DIESEL,863,R$/l,3.448,...,4.220,0.401,0.045,3.047,0.097,2.8599,3.3481,0.032,6,2019
4259,106751,106751,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,ÓLEO DIESEL S10,1277,R$/l,3.574,...,4.320,0.451,0.047,3.123,0.111,2.885,3.491,0.036,6,2019
4260,106778,106778,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,GASOLINA COMUM,1644,R$/l,4.142,...,4.999,0.429,0.044,3.713,0.097,3.52,4.0892,0.026,6,2019
4261,106805,106805,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,GLP,1017,R$/13Kg,66.893,...,89.900,17.679,0.091,49.214,5.071,40.19,72,0.103,6,2019


Alternativamente, poderíamos usar o método `query` para fazermos tal seleção. Porém, isso não é possível especificamente para esse caso, pois o rótulo da coluna 'PREÇO MÉDIO REVENDA' possui caracteres inválidos para o método (cedilha, acentos)

In [ ]:
# Não funciona
# data.query('ESTADO' == "RIO DE JANEIRO" and PREÇO MÉDIO REVENDA > 2')

In [ ]:
# data.query('ESTADO' == "RIO DE JANEIRO" or ESTADO == "SAO PAULO"')

**Aprofundando mais ainda**

A primeira comparação `(data['ESTADO'] == 'RIO DE JANEIRO')` checa, linha a linha (amostra a amostra) do DataFrame, quais são aquelas cujo o estado é RIO DE JANEIRO. Nenhuma averiguação de preços é feita nesse momento. Como resultado, temos uma Series de booleans que responde **apenas** a essa "pergunta" feita.

A segunda comparação `(data['PREÇO MÉDIO REVENDA'] > 2)` checa, linha a linha (amostra a amostra) do DataFrame, quais são os registro cujo preço do combustível é maior que 2 reais. Note que essa comparação checará os postos de **TODOS os estados**. Como resultado, temos uma Series de booleans que responde **apenas** a essa "pergunta" feita.

Por fim, as duas "perguntas" são unidas pelo AND (&) que retorna a "pergunta completa" que fizemos.

Alguns podem argumentar que tal abordagem é **ineficiente**, uma vez que, para cada condição ("pergunta"), estamos varrendo todas as linhas do DataFrame. <br/>
O Pandas _tenta otimizar_ isso ao máximo por de trás dos panos. Mas, de fato, de tivermos um dataset **muito grande** (centenas de milhares de linhas), tal abordagem se tornará _lenta_.

Assim, poderíamos fazer filtragem com múltiplos condicionais em partes:

In [ ]:
selecao_1 = data['ESTADO'] == 'RIO DE JANEIRO'
postos_rj = data[selecao_1]
postos_rj


,Unnamed: 0.1,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
22,22,22,2004-05-09,2004-05-15,SUDESTE,RIO DE JANEIRO,ETANOL HIDRATADO,952,R$/l,1.098,...,1.890,0.312,0.140,0.786,0.173,0.326,1.263,0.22,5,2004
49,49,49,2004-05-16,2004-05-22,SUDESTE,RIO DE JANEIRO,ETANOL HIDRATADO,948,R$/l,1.092,...,1.499,0.284,0.144,0.808,0.175,0.3257,1.2921,0.217,5,2004
76,76,76,2004-05-23,2004-05-29,SUDESTE,RIO DE JANEIRO,ETANOL HIDRATADO,973,R$/l,1.103,...,1.589,0.281,0.127,0.822,0.166,0.38909,1.1859,0.202,5,2004
103,103,103,2004-05-30,2004-06-05,SUDESTE,RIO DE JANEIRO,ETANOL HIDRATADO,955,R$/l,1.149,...,1.599,0.3,0.136,0.849,0.188,0.3453,1.3667,0.221,5,2004
130,130,130,2004-06-06,2004-06-12,SUDESTE,RIO DE JANEIRO,ETANOL HIDRATADO,966,R$/l,1.165,...,1.599,0.254,0.132,0.911,0.198,0.4,1.3667,0.217,6,2004
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
106718,106718,106718,2019-06-23,2019-06-29,SUDESTE,RIO DE JANEIRO,ÓLEO DIESEL,212,R$/l,3.569,...,4.059,0.467,0.047,3.102,0.119,2.886,3.5162,0.038,6,2019
106745,106745,106745,2019-06-23,2019-06-29,SUDESTE,RIO DE JANEIRO,ÓLEO DIESEL S10,335,R$/l,3.697,...,4.389,0.5,0.051,3.197,0.127,2.9772,3.5388,0.04,6,2019
106772,106772,106772,2019-06-23,2019-06-29,SUDESTE,RIO DE JANEIRO,GASOLINA COMUM,422,R$/l,4.928,...,5.859,0.538,0.033,4.39,0.097,4.2,4.618,0.022,6,2019
106799,106799,106799,2019-06-23,2019-06-29,SUDESTE,RIO DE JANEIRO,GLP,221,R$/13Kg,65.010,...,79.000,17.475,0.089,47.535,3.934,39,58.2,0.083,6,2019


In [ ]:
selecao_2 = postos_rj['PREÇO MÉDIO REVENDA'] > 2
selecao_2

,PREÇO MÉDIO REVENDA
22,False
49,False
76,False
103,False
130,False
...,...
106718,True
106745,True
106772,True
106799,True


In [ ]:
postos_rj_preco_maior_que_2 = postos_rj[selecao_2]
postos_rj_preco_maior_que_2

,Unnamed: 0.1,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
2501,2501,2501,2006-02-26,2006-03-04,SUDESTE,RIO DE JANEIRO,ETANOL HIDRATADO,881,R$/l,2.028,...,2.550,0.222,0.073,1.806,0.11,1.31,2.1358,0.061,2,2006
2528,2528,2528,2006-03-05,2006-03-11,SUDESTE,RIO DE JANEIRO,ETANOL HIDRATADO,889,R$/l,2.125,...,2.550,0.286,0.070,1.839,0.117,1.438,2.2801,0.064,3,2006
2555,2555,2555,2006-03-12,2006-03-18,SUDESTE,RIO DE JANEIRO,ETANOL HIDRATADO,902,R$/l,2.163,...,2.599,0.268,0.064,1.895,0.134,1.43,2.3454,0.071,3,2006
2582,2582,2582,2006-03-19,2006-03-25,SUDESTE,RIO DE JANEIRO,ETANOL HIDRATADO,898,R$/l,2.166,...,2.568,0.231,0.067,1.935,0.14,1.438,2.2925,0.072,3,2006
2609,2609,2609,2006-03-26,2006-04-01,SUDESTE,RIO DE JANEIRO,ETANOL HIDRATADO,890,R$/l,2.172,...,2.550,0.209,0.066,1.963,0.133,1.438,2.2513,0.068,3,2006
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
106718,106718,106718,2019-06-23,2019-06-29,SUDESTE,RIO DE JANEIRO,ÓLEO DIESEL,212,R$/l,3.569,...,4.059,0.467,0.047,3.102,0.119,2.886,3.5162,0.038,6,2019
106745,106745,106745,2019-06-23,2019-06-29,SUDESTE,RIO DE JANEIRO,ÓLEO DIESEL S10,335,R$/l,3.697,...,4.389,0.5,0.051,3.197,0.127,2.9772,3.5388,0.04,6,2019
106772,106772,106772,2019-06-23,2019-06-29,SUDESTE,RIO DE JANEIRO,GASOLINA COMUM,422,R$/l,4.928,...,5.859,0.538,0.033,4.39,0.097,4.2,4.618,0.022,6,2019
106799,106799,106799,2019-06-23,2019-06-29,SUDESTE,RIO DE JANEIRO,GLP,221,R$/13Kg,65.010,...,79.000,17.475,0.089,47.535,3.934,39,58.2,0.083,6,2019


#### **Selecionando registros de postos de São Paulo ou do Rio de Janeiro com Gasolina Comum acima de 2 reais**

Podemos fazer a solução do "jeito mais lento", percorrendo o DataFrame inteiro _múltiplas vezes_:

In [19]:
data.head()

,",Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,PREÇO MÍNIMO REVENDA,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO"
0,"0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRIT..."
1,"1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,E..."
2,"2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GR..."
3,"3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GR..."
4,"4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETA..."


In [20]:
selecao_1 = (data['ESTADO'] == 'SAO PAULO') | (data['ESTADO'] == 'RIO DE JANEIRO')
selecao_2 = (data['PRODUTO'] == 'GASOLINA COMUM')
selecao_3 = (data['PREÇO MÉDIO REVENDA'] > 2)

selecao_final = selecao_1 & selecao_2 & selecao_3

KeyError: 'ESTADO'

In [21]:
selecao_final

NameError: name 'selecao_final' is not defined

In [22]:
data_filtrado = data[selecao_final]
data_filtrado

NameError: name 'selecao_final' is not defined

<br/>

Alternativamente:

In [23]:
selecao_1 = (data['ESTADO'] == 'SAO PAULO') | (data['ESTADO'] == 'RIO DE JANEIRO')
postos_sp_rj = data[selecao_1]
postos_sp_rj

KeyError: 'ESTADO'

In [24]:
selecao_2 = (postos_sp_rj['PRODUTO'] == 'GASOLINA COMUM')
postos_sp_rj_gasolina = postos_sp_rj[selecao_2]
postos_sp_rj_gasolina

NameError: name 'postos_sp_rj' is not defined

In [25]:
selecao_3 = (postos_sp_rj_gasolina['PREÇO MÉDIO REVENDA'] > 2)
postos_sp_rj_maior_que_2 = postos_sp_rj_gasolina[selecao_3]
postos_sp_rj_maior_que_2


NameError: name 'postos_sp_rj_gasolina' is not defined

#### **Selecionando registros dos anos de 2008, 2010 e 2012**

**ALTERNATIVA 1**

In [26]:
selecao = (data['ANO'] == 2008) | (data['ANO'] == 2009) | (data['ANO'] == 2010) | (data['ANO'] == 2012)
display(data[selecao].head())


KeyError: 'ANO'

**ALTERNATIVA 2**

In [ ]:
lista_de_anos = [2008, 2010, 2012]

In [27]:
lista_de_anos = [2008, 2010, 2012]
selecao = data['ANO'].isin(lista_de_anos)
display(data[selecao].head())

KeyError: 'ANO'

In [28]:
print(data.columns)

Index([',Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,PREÇO MÍNIMO REVENDA,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO'], dtype='object')


### **Iterando com DataFrames**

#### For-each `DataFrame.iterrows()` (LENTO ==> apenas indicado para iterar pequenos conjunto de dados)

In [29]:
for index, row in data.head(10).iterrows():
  print(f'indice{index} ==> {row["ESTADO"]}')

KeyError: 'ESTADO'

<h2>2. Preparação dos dados</h2>
<hr/>

In [30]:
data

,",Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,PREÇO MÍNIMO REVENDA,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO"
0,"0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRIT..."
1,"1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,E..."
2,"2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GR..."
3,"3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GR..."
4,"4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETA..."
...,...
106818,"106818,106818,2019-06-23,2019-06-29,NORDESTE,R..."
106819,"106819,106819,2019-06-23,2019-06-29,SUL,RIO GR..."
106820,"106820,106820,2019-06-23,2019-06-29,SUL,SANTA ..."
106821,"106821,106821,2019-06-23,2019-06-29,SUDESTE,SA..."


In [31]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 106823 entries, 0 to 106822
Data columns (total 1 columns):
 #   Column                                                                                                                                                                                                                                                                                                                                                                                    Non-Null Count   Dtype 
---  ------                                                                                                                                                                                                                                                                                                                                                                                    --------------   ----- 
 0   ,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POS

### 2.1 Removendo amostras com valores vazios (null / nan) no dataset

De um total de 106823 observações, **não há valores null** / nan para nenhum atributo. Mas, veremos que não é bem assim neste caso específico.<br/><br/>

### 2.2 Conversão de tipos de atributos

O pandas automaticamente reconhece os tipos de dados de cada coluna. <br/>
Porém, existem alguns atributos que estão com seus tipos errados: P. ex., "PREÇO MÉDIO DISTRIBUIÇÃO" deveria ser ```float64``` e não ```object```.<br/>
Nestes casos, muito provavelmente algumas amostras têm um string ao invés de um número para tais atributos. <br/>

Os atributos *"DATA INICIAL"* e *"DATA FINAL"* deveriam ser do tipo `datetime`.

Em outros casos, alguns **atributos categóricos** são ```objects```, mas poderiam ter o tipo ```category```, que é um tipo especial do pandas. <br/>
Este tipo é necessário para se utilizar algumas funções específicas do pandas. <br/>
**Não** converteremos para este tipo por ora.

In [32]:
 data_pre = data.copy()


#### **Datas**
Como os atributos de data do datset já estão em um formato de data aceitável (YYYY-MM-DD), não precisamos forçar nenhuma conversão nesse sentido.

In [35]:

data_pre['DATA INICIAL'] = pd.to_datetime(data_pre['DATA INICIAL'])
data_pre['DATA FINAL'] = pd.to_datetime(data_pre['DATA FINAL'])
data_pre.info()

KeyError: 'DATA INICIAL'

#### **Dados Numéricos**

In [36]:

for atributo in ['MARGEM MÉDIA REVENDA', 'PREÇO MÉDIO DISTRIBUIÇÃO', 'DESVIO PADRÃO DISTRIBUIÇÃO', 'PREÇO MÍNIMO DISTRIBUIÇÃO', 'PREÇO MÁXIMO DISTRIBUIÇÃO', 'COEF DE VARIAÇÃO DISTRIBUIÇÃO']:
 data_pre[atributo] = pd.to_numeric(data_pre[atributo], errors='coerce')

KeyError: 'MARGEM MÉDIA REVENDA'

In [38]:
data_pre.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 106823 entries, 0 to 106822
Data columns (total 1 columns):
 #   Column                                                                                                                                                                                                                                                                                                                                                                                    Non-Null Count   Dtype 
---  ------                                                                                                                                                                                                                                                                                                                                                                                    --------------   ----- 
 0   ,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POS

<br/>

Note que temos vários valores ***null*** agora **após a *conversão de tipos***. Vamos checá-los com mais cuidado nos dados originais e preprocessados.

### 2.3 Limpeza de dados

In [39]:
mask = data_pre['PREÇO MÉDIO DISTRIBUIÇÃO'].isnull()
data_pre[mask]


KeyError: 'PREÇO MÉDIO DISTRIBUIÇÃO'

In [40]:
data_pre[mask]

NameError: name 'mask' is not defined

Várias amostras possuem a _string '-'_ em algumas colunas ao invés de um número de fato. Ou seja, não há aferições destes atributos para estas amostras. <br/>

<br/>

Poderíamos **preencher os valores NaN com um valor padrão**. Para isso, basta usar o método `.fillna`.

In [42]:
# Retorna uma cópia do data fram `data_pre` com todos os valores NaN de todas as colunas agora preenchidos com 0.
# Para alterar o próprio data frame, use o argumento `inplace=True`.

data_pre_fill = data_pre.fillna(0)
data_pre_fill


,",Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,PREÇO MÍNIMO REVENDA,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO"
0,"0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRIT..."
1,"1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,E..."
2,"2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GR..."
3,"3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GR..."
4,"4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETA..."
...,...
106818,"106818,106818,2019-06-23,2019-06-29,NORDESTE,R..."
106819,"106819,106819,2019-06-23,2019-06-29,SUL,RIO GR..."
106820,"106820,106820,2019-06-23,2019-06-29,SUL,SANTA ..."
106821,"106821,106821,2019-06-23,2019-06-29,SUDESTE,SA..."


In [43]:
data_pre_fill[mask]  # valores que antes eram NaN, agora são 0


NameError: name 'mask' is not defined

In [46]:
# retorna uma cópia do data frame `data_pre` com todos os valores NaN das colunas:
# 'PREÇO MÉDIO DISTRIBUIÇÃO', 'DESVIO PADRÃO DISTRIBUIÇÃO', 'PREÇO MÍNIMO DISTRIBUIÇÃO' e 'PREÇO MÁXIMO DISTRIBUIÇÃO', respectivamente, com os valores: 10, 20, 30, 'vazio'.
data_pre_fill = data_pre.fillna(value={
    'PREÇO MÉDIO DISTRIBUIÇÃO': 10,
    'DESVIO PADRÃO DISTRIBUIÇÃO': 20,
    'PREÇO MÍNIMO DISTRIBUIÇÃO': 30,
    'PREÇO MÁXIMO DISTRIBUIÇÃO': 'vazio'
})

data_pre_fill[mask]



NameError: name 'mask' is not defined

<br/>

Por mais que a função `fillna` seja interessante e útil em muitos casos, no problema em questão estamos interessados em analisar precisamente, p. ex., o **'PREÇO MÉDIO DISTRIBUIÇÃO'**.<br/>
A fim de não termos valores _sintéticos_ gerados pelo `fillna`, que possam atrapalhar nossa análise, iremos **remover (drop) todas as amostras que possuem qualquer valor NaN** para quaisquer atributos/colunas. <br/>

Para isso, basta utilizarmos o método `dropna`.

In [47]:
# remove, no próprio dataframe, todas as linhas/registros com valores NaN (vazios) em quaisquer colunas/atributos.
data_pre.dropna(inplace=True)
data_pre.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 106823 entries, 0 to 106822
Data columns (total 1 columns):
 #   Column                                                                                                                                                                                                                                                                                                                                                                                    Non-Null Count   Dtype 
---  ------                                                                                                                                                                                                                                                                                                                                                                                    --------------   ----- 
 0   ,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POS

Nosso data frame agora, após essa _limpeza_, ficou mais enxuto, contentdo 103392 registros frente aos 106823 registros originais. <br/>

Essas são apenas algumas das possíveis _técnicas de limpeza de dados_. Outras estratégias, p. ex., **confiam na detecção de outliers**, que não veremos neste curso.

#### **Salvando o Dataset Preprocessado**

In [48]:
data_pre.to_csv('/content/drive/MyDrive/Trilha de Dados/GasPricesinBrazil_2004-2019 (1).csv', index=False\)

SyntaxError: unexpected character after line continuation character (ipython-input-560306730.py, line 1)

<h2>3. Estatísticas Descritivas</h2>
<hr/>

O Pandas fornecem algumas funções/métodos ue computam certas estatísticas descritivas.

`describe`: exibe várias **estatísticas descritivas** para os _atributos_ de um dataframe ou para uma _Series_.

In [49]:
data_final = pd.read_csv('/content/drive/MyDrive/Trilha de Dados/GasPricesinBrazil_2004-2019 (1).csv')
data_final

,Unnamed: 0.1,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
4,4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,...,1.400,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
106818,106818,106818,2019-06-23,2019-06-29,NORDESTE,RIO GRANDE DO NORTE,GNV,7,R$/m3,3.574,...,3.690,0.818,0.018,2.756,0,2.756,2.756,0,6,2019
106819,106819,106819,2019-06-23,2019-06-29,SUL,RIO GRANDE DO SUL,GNV,23,R$/m3,3.401,...,3.789,0.95,0.038,2.451,0.402,1.9842,2.8661,0.164,6,2019
106820,106820,106820,2019-06-23,2019-06-29,SUL,SANTA CATARINA,GNV,24,R$/m3,2.912,...,3.499,0.914,0.065,1.998,0,1.9981,1.9981,0,6,2019
106821,106821,106821,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,GNV,52,R$/m3,3.020,...,3.490,0.646,0.076,2.374,0.165,2.0179,2.5093,0.07,6,2019


In [50]:
data_final.describe()

,Unnamed: 0.1,Unnamed: 0,NÚMERO DE POSTOS PESQUISADOS,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,PREÇO MÍNIMO REVENDA,PREÇO MÁXIMO REVENDA,COEF DE VARIAÇÃO REVENDA,MÊS,ANO
count,106823.00000,106823.00000,106823.000000,106823.000000,106823.000000,106823.000000,106823.000000,106823.000000,106823.000000,106823.000000
mean,53411.00000,53411.00000,233.770976,10.870780,0.657227,9.416038,12.537731,0.045418,6.484006,2011.802271
std,30837.28824,30837.28824,403.282519,17.752533,1.349169,15.092193,20.841497,0.025066,3.443391,4.391530
min,0.00000,0.00000,1.000000,0.766000,0.000000,0.590000,0.999000,0.000000,1.000000,2004.000000
25%,26705.50000,26705.50000,42.000000,2.072000,0.072000,1.900000,2.310000,0.029000,4.000000,2008.000000
50%,53411.00000,53411.00000,104.000000,2.718000,0.114000,2.510000,2.999000,0.041000,6.000000,2012.000000
75%,80116.50000,80116.50000,243.000000,3.752000,0.191000,3.499000,4.150000,0.058000,9.000000,2016.000000
max,106822.00000,106822.00000,4167.000000,99.357000,10.748000,90.000000,120.000000,0.395000,12.000000,2019.000000


In [51]:
data_final['PREÇO MÉDIO REVENDA'].describe()

,PREÇO MÉDIO REVENDA
count,106823.000000
mean,10.870780
std,17.752533
min,0.766000
25%,2.072000
50%,2.718000
75%,3.752000
max,99.357000


In [52]:
data_final.describe()['PREÇO MÉDIO REVENDA']

,PREÇO MÉDIO REVENDA
count,106823.000000
mean,10.870780
std,17.752533
min,0.766000
25%,2.072000
50%,2.718000
75%,3.752000
max,99.357000


In [53]:
stats = data_final.describe()
stats


,Unnamed: 0.1,Unnamed: 0,NÚMERO DE POSTOS PESQUISADOS,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,PREÇO MÍNIMO REVENDA,PREÇO MÁXIMO REVENDA,COEF DE VARIAÇÃO REVENDA,MÊS,ANO
count,106823.00000,106823.00000,106823.000000,106823.000000,106823.000000,106823.000000,106823.000000,106823.000000,106823.000000,106823.000000
mean,53411.00000,53411.00000,233.770976,10.870780,0.657227,9.416038,12.537731,0.045418,6.484006,2011.802271
std,30837.28824,30837.28824,403.282519,17.752533,1.349169,15.092193,20.841497,0.025066,3.443391,4.391530
min,0.00000,0.00000,1.000000,0.766000,0.000000,0.590000,0.999000,0.000000,1.000000,2004.000000
25%,26705.50000,26705.50000,42.000000,2.072000,0.072000,1.900000,2.310000,0.029000,4.000000,2008.000000
50%,53411.00000,53411.00000,104.000000,2.718000,0.114000,2.510000,2.999000,0.041000,6.000000,2012.000000
75%,80116.50000,80116.50000,243.000000,3.752000,0.191000,3.499000,4.150000,0.058000,9.000000,2016.000000
max,106822.00000,106822.00000,4167.000000,99.357000,10.748000,90.000000,120.000000,0.395000,12.000000,2019.000000


In [54]:
stats[['PREÇO MÉDIO REVENDA', 'PREÇO MÁXIMO REVENDA', 'PREÇO MÉDIO DISTRIBUIÇÃO']]


KeyError: "['PREÇO MÉDIO DISTRIBUIÇÃO'] not in index"

In [55]:
# ALTERNATIVA MAIS EFICIENTE
# apenas computa as estatísticas descritivas para 3 atributos
data_final[['PREÇO MÉDIO REVENDA', 'PREÇO MÁXIMO REVENDA', 'PREÇO MÉDIO DISTRIBUIÇÃO']].describe()


,PREÇO MÉDIO REVENDA,PREÇO MÁXIMO REVENDA
count,106823.000000,106823.000000
mean,10.870780,12.537731
std,17.752533,20.841497
min,0.766000,0.999000
25%,2.072000,2.310000
50%,2.718000,2.999000
75%,3.752000,4.150000
max,99.357000,120.000000


<br/>

Como o resultado do `describe` de um _dataframe_ é outro _dataframe_, podemos filtrar apenas algumas colunas.

**Acessando apenas algumas estatísticas**

In [56]:
stats

,Unnamed: 0.1,Unnamed: 0,NÚMERO DE POSTOS PESQUISADOS,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,PREÇO MÍNIMO REVENDA,PREÇO MÁXIMO REVENDA,COEF DE VARIAÇÃO REVENDA,MÊS,ANO
count,106823.00000,106823.00000,106823.000000,106823.000000,106823.000000,106823.000000,106823.000000,106823.000000,106823.000000,106823.000000
mean,53411.00000,53411.00000,233.770976,10.870780,0.657227,9.416038,12.537731,0.045418,6.484006,2011.802271
std,30837.28824,30837.28824,403.282519,17.752533,1.349169,15.092193,20.841497,0.025066,3.443391,4.391530
min,0.00000,0.00000,1.000000,0.766000,0.000000,0.590000,0.999000,0.000000,1.000000,2004.000000
25%,26705.50000,26705.50000,42.000000,2.072000,0.072000,1.900000,2.310000,0.029000,4.000000,2008.000000
50%,53411.00000,53411.00000,104.000000,2.718000,0.114000,2.510000,2.999000,0.041000,6.000000,2012.000000
75%,80116.50000,80116.50000,243.000000,3.752000,0.191000,3.499000,4.150000,0.058000,9.000000,2016.000000
max,106822.00000,106822.00000,4167.000000,99.357000,10.748000,90.000000,120.000000,0.395000,12.000000,2019.000000


In [57]:
stats.loc[['min', 'max', 'mean']]

,Unnamed: 0.1,Unnamed: 0,NÚMERO DE POSTOS PESQUISADOS,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,PREÇO MÍNIMO REVENDA,PREÇO MÁXIMO REVENDA,COEF DE VARIAÇÃO REVENDA,MÊS,ANO
min,0.0,0.0,1.000000,0.76600,0.000000,0.590000,0.999000,0.000000,1.000000,2004.000000
max,106822.0,106822.0,4167.000000,99.35700,10.748000,90.000000,120.000000,0.395000,12.000000,2019.000000
mean,53411.0,53411.0,233.770976,10.87078,0.657227,9.416038,12.537731,0.045418,6.484006,2011.802271


In [58]:
stats.loc[['min', 'max', 'mean'], ['PREÇO MÉDIO REVENDA', 'PREÇO MÉDIO DISTRIBUIÇÃO']]

KeyError: "['PREÇO MÉDIO DISTRIBUIÇÃO'] not in index"

<br/>

`mean`, `std`, `min`, etc: cada uma das estatísticas do `describe` podem ser computadas individualmente:

In [59]:
data_final.head()

,Unnamed: 0.1,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
4,4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,...,1.400,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004


#### Qual é o menor preço mínimo de revenda?

In [60]:
data_final['PREÇO MÍNIMO REVENDA'].min()

0.59

#### Qual é a média e desvio padrão dos preços mínimos de revenda?

In [61]:

mean = data_final['PREÇO MÍNIMO REVENDA'].mean()
std = data_final['PREÇO MÍNIMO REVENDA'].std()

print(f'A média dos preços mínimos de revenda é {mean:.2f} +- {std:.2f}')

A média dos preços mínimos de revenda é 9.42 +- 15.09


#### Quais são os estados considerados?

In [62]:
data_final['ESTADO']


,ESTADO
0,DISTRITO FEDERAL
1,GOIAS
2,MATO GROSSO
3,MATO GROSSO DO SUL
4,ALAGOAS
...,...
106818,RIO GRANDE DO NORTE
106819,RIO GRANDE DO SUL
106820,SANTA CATARINA
106821,SAO PAULO


In [63]:
data_final['ESTADO'].unique()

array(['DISTRITO FEDERAL', 'GOIAS', 'MATO GROSSO', 'MATO GROSSO DO SUL',
       'ALAGOAS', 'BAHIA', 'CEARA', 'MARANHAO', 'PARAIBA', 'PERNAMBUCO',
       'PIAUI', 'RIO GRANDE DO NORTE', 'SERGIPE', 'ACRE', 'AMAPA',
       'AMAZONAS', 'PARA', 'RONDONIA', 'RORAIMA', 'TOCANTINS',
       'ESPIRITO SANTO', 'MINAS GERAIS', 'RIO DE JANEIRO', 'SAO PAULO',
       'PARANA', 'RIO GRANDE DO SUL', 'SANTA CATARINA'], dtype=object)

In [64]:
sorted(data_final['ESTADO'].unique())

['ACRE',
 'ALAGOAS',
 'AMAPA',
 'AMAZONAS',
 'BAHIA',
 'CEARA',
 'DISTRITO FEDERAL',
 'ESPIRITO SANTO',
 'GOIAS',
 'MARANHAO',
 'MATO GROSSO',
 'MATO GROSSO DO SUL',
 'MINAS GERAIS',
 'PARA',
 'PARAIBA',
 'PARANA',
 'PERNAMBUCO',
 'PIAUI',
 'RIO DE JANEIRO',
 'RIO GRANDE DO NORTE',
 'RIO GRANDE DO SUL',
 'RONDONIA',
 'RORAIMA',
 'SANTA CATARINA',
 'SAO PAULO',
 'SERGIPE',
 'TOCANTINS']

#### Quantos registros (aferições) cada estado possui?

`.value_counts()`:  Conta a frequência dos valores de uma dada variável (de preferência, _categórica_).

In [65]:
data_final['ESTADO'].value_counts()  # retorna em ordem decrescente, a quantidade de registros/linhas para cada estado

,count
ESTADO,
PARAIBA,4263
BAHIA,4263
CEARA,4263
RIO GRANDE DO SUL,4263
RIO DE JANEIRO,4263
SAO PAULO,4263
PERNAMBUCO,4262
RIO GRANDE DO NORTE,4262
SANTA CATARINA,4262


In [66]:
data_final['ESTADO'].value_counts()  # retorna em ordem decrescente, a quantidade de registros/linhas para cada estado


,count
ESTADO,
PARAIBA,4263
BAHIA,4263
CEARA,4263
RIO GRANDE DO SUL,4263
RIO DE JANEIRO,4263
SAO PAULO,4263
PERNAMBUCO,4262
RIO GRANDE DO NORTE,4262
SANTA CATARINA,4262


<h2>4. Executando funções para cada item de um DataFrame ou Series</h2>
<hr/>

Uma alternativa ao `for-loop` que vimos anteriormente e que é _lento_, é usarmos _funções próprias do pandas_ que **aplicam/mapeiam uma dada função a todos os elementos de um DataFrame ou Series**, retornando novos elementos "transformados".


<img src='./imagens/apply_map_applymap.png' width=300/>


Fonte: https://towardsdatascience.com/introduction-to-pandas-apply-applymap-and-map-5d3e044e93ff

In [ ]:
df = pd.DataFrame({ 'A': [1, 2, 3, 4],
                    'B': [10, 20, 30, 40],
                    'C': [100, 200, 300, 400]},
                     index=['Linha 1', 'Linha 2', 'Linha 3', 'Linha 4'])

In [67]:
df

NameError: name 'df' is not defined

`apply()`: usado para aplicar uma função ao longo de um eixo de um DataFrame ou em valores de uma Series.

<img src='./imagens/pandas_axis.jpg' width=500/>

Fonte: https://www.allthesnippets.com/browse/pandas/df_axis.html

<img src='./imagens/apply_axis_1.png' width=250/>

<img src='./imagens/apply_axis_0.png' width=250/>

##### Usando `lambda` functions

<img src='./imagens/apply_axis_1_mean.png' width=350/>

<br/>

`applymap()`: usado para aplicar uma função para **cada elemento** (_element-wise_) de um DataFrame.

In [ ]:
df = pd.DataFrame({ 'A': [1, 2, 3, 4],
                    'B': [10, 20, 30, 40],
                    'C': [100, 200, 300, 400]},
                     index=['Linha 1', 'Linha 2', 'Linha 3', 'Linha 4'])
df

<br/>

`map()`: usado para aplicar uma função para **cada elemento** (_element-wise_) de uma _Series_.

In [69]:
def nossa_soma(series):
    return series.sum()  # retorna a soma de todos os valores de uma series


In [68]:
# aplica a função soma para cada linha do dataframe
df['SOMA(A, B, C)'] = df.apply(nossa_soma, axis=1)
df


NameError: name 'df' is not defined

In [70]:
# aplica a função soma para cada coluna do dataframe
df.loc['Linha 5'] = df.apply(nossa_soma, axis=0)
df



NameError: name 'df' is not defined

In [71]:
df['MEDIA(A, B, C)'] = df[['A', 'B', 'C']].apply(lambda series: series.mean(), axis=1)
df

NameError: name 'df' is not defined

In [72]:
# plica a lambda function abaixo para cada elemento da coluna
df['C * 2'] = df['C'].apply(lambda x: x * 2)
df

NameError: name 'df' is not defined

In [73]:
df['A * 2'] = df['A'] * 2
df

NameError: name 'df' is not defined

<h2>5. Agrupamento</h2>
<hr/>

`groupby`: Usado para criar **grupo de elementos** (e.x., baseado nos valores de um atributo). <br/>
**Funções** podem então ser aplicadas para os _elementos de cada grupo_, de modo que os **resultados de cada grupo são combinados**.

In [74]:

data_final.head()

,Unnamed: 0.1,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
4,4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,...,1.400,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004


In [75]:
# agrupa as linhas da tablea de acordo com suas respectivas regiões
grupos = data_final.groupby('REGIÃO')
grupos


In [76]:
# retorna os grupos obtidos pelo groupby
grupos.groups


{'CENTRO OESTE': [0, 1, 2, 3, 27, 28, 29, 30, 54, 55, 56, 57, 81, 82, 83, 84, 108, 109, 110, 111, 135, 136, 137, 138, 162, 163, 164, 165, 189, 190, 191, 192, 216, 217, 218, 219, 243, 244, 245, 246, 270, 271, 272, 273, 297, 298, 299, 300, 324, 325, 326, 327, 351, 352, 353, 354, 378, 379, 380, 381, 405, 406, 407, 408, 432, 433, 434, 435, 459, 460, 461, 462, 486, 487, 488, 489, 513, 514, 515, 516, 540, 541, 542, 543, 567, 568, 569, 570, 594, 595, 596, 597, 621, 622, 623, 624, 648, 649, 650, 651, ...], 'NORDESTE': [4, 5, 6, 7, 8, 9, 10, 11, 12, 31, 32, 33, 34, 35, 36, 37, 38, 39, 58, 59, 60, 61, 62, 63, 64, 65, 66, 85, 86, 87, 88, 89, 90, 91, 92, 93, 112, 113, 114, 115, 116, 117, 118, 119, 120, 139, 140, 141, 142, 143, 144, 145, 146, 147, 166, 167, 168, 169, 170, 171, 172, 173, 174, 193, 194, 195, 196, 197, 198, 199, 200, 201, 220, 221, 222, 223, 224, 225, 226, 227, 228, 247, 248, 249, 250, 251, 252, 253, 254, 255, 274, 275, 276, 277, 278, 279, 280, 281, 282, 301, ...], 'NORTE': [13, 14, 1

In [77]:
# retorna os índices das linhas/observações de cada grupo
grupos.indices



{'CENTRO OESTE': array([     0,      1,      2, ..., 106791, 106792, 106812]),
 'NORDESTE': array([     4,      5,      6, ..., 106816, 106818, 106822]),
 'NORTE': array([    13,     14,     15, ..., 106802, 106803, 106807]),
 'SUDESTE': array([    20,     21,     22, ..., 106813, 106817, 106821]),
 'SUL': array([    24,     25,     26, ..., 106815, 106819, 106820])}

In [78]:
# retorna um dataframe apenas com as observações do grupo 'CENTRO OESTE'
grupos.get_group('CENTRO OESTE')

,Unnamed: 0.1,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
27,27,27,2004-05-16,2004-05-22,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,144,R$/l,1.271,...,1.300,0.448,0.031,0.823,0.111,0.4094,1.1931,0.135,5,2004
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
106787,106787,106787,2019-06-23,2019-06-29,CENTRO OESTE,DISTRITO FEDERAL,GLP,42,R$/13Kg,73.452,...,95.000,21.509,0.066,51.943,7.877,44,62.82,0.152,6,2019
106789,106789,106789,2019-06-23,2019-06-29,CENTRO OESTE,GOIAS,GLP,209,R$/13Kg,71.968,...,85.000,15.044,0.075,56.924,4.285,45.57,61.6,0.075,6,2019
106791,106791,106791,2019-06-23,2019-06-29,CENTRO OESTE,MATO GROSSO,GLP,95,R$/13Kg,96.691,...,115.000,19.789,0.071,76.902,8.846,61.15,90,0.115,6,2019
106792,106792,106792,2019-06-23,2019-06-29,CENTRO OESTE,MATO GROSSO DO SUL,GLP,115,R$/13Kg,72.297,...,99.000,16.707,0.111,55.59,5.109,47,72,0.092,6,2019


In [79]:
# descreva para nós algumas estatística descritivas para as observações de cada grupo
grupos.describe()


Unnamed: 0.1                                              \
                    count          mean           std   min       25%   
REGIÃO                                                                  
CENTRO OESTE      15429.0  53417.968955  30938.517860   0.0  26021.00   
NORDESTE          36869.0  53538.276248  30636.581838   4.0  27645.00   
NORTE             24702.0  52805.662214  31594.475085  13.0  23848.25   
SUDESTE           17047.0  53749.005104  30386.889753  20.0  28766.50   
SUL               12776.0  53754.692001  30391.714995  24.0  28743.75   

                                          Unnamed: 0                ...   MÊS  \
                  50%       75%       max      count          mean  ...   75%   
REGIÃO                                                              ...         
CENTRO OESTE  54240.0  80018.00  106812.0    15429.0  53417.968955  ...  9.00   
NORDESTE      53027.0  80098.00  106822.0    36869.0  53538.276248  ...  9.00   
NORTE         54009.5  80299.50  106807.0    24702.0  52805.662214  ...  9.00   
SUDESTE       52829.0  80048.50  106821.0    17047.0  53749.005104  ...  9.00   
SUL           52859.5  80059.75  106820.0    12776.0  53754.692001  ...  9.25   

                        ANO                                                 \
               max    count         mean       std     min     25%     50%   
REGIÃO                                                                       
CENTRO OESTE  12.0  15429.0  2011.862532  4.352302  2004.0  2008.0  2012.0   
NORDESTE      12.0  36869.0  2011.777699  4.404509  2004.0  2008.0  2012.0   
NORTE         12.0  24702.0  2011.832524  4.386504  2004.0  2008.0  2012.0   
SUDESTE       12.0  17047.0  2011.775796  4.399684  2004.0  2008.0  2012.0   
SUL           12.0  12776.0  2011.777239  4.399609  2004.0  2008.0  2012.0   

                              
                 75%     max  
REGIÃO                        
CENTRO OESTE  2016.0  2019.0  
NORDESTE      2016.0  2019.0  
NORTE         2016.0  2019.0  
SUDESTE       2016.0  2019.0  
SUL           2016.0  2019.0  

[5 rows x 80 columns]

In [80]:
grupos.mean()


TypeError: agg function failed [how->mean,dtype->object]

In [81]:
grupos.min()


,Unnamed: 0.1,Unnamed: 0,DATA INICIAL,DATA FINAL,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
REGIÃO,,,,,,,,,,,,,,,,,,,,,
CENTRO OESTE,0,0,2004-05-09,2004-05-15,DISTRITO FEDERAL,ETANOL HIDRATADO,1,R$/13Kg,1.097,0.0,...,1.099,-,0.0,-,-,-,-,-,1,2004
NORDESTE,4,4,2004-05-09,2004-05-15,ALAGOAS,ETANOL HIDRATADO,1,R$/13Kg,0.975,0.0,...,0.999,-,0.0,-,-,-,-,-,1,2004
NORTE,13,13,2004-05-09,2004-05-15,ACRE,ETANOL HIDRATADO,1,R$/13Kg,1.219,0.0,...,1.219,-,0.0,-,-,-,-,-,1,2004
SUDESTE,20,20,2004-05-09,2004-05-15,ESPIRITO SANTO,ETANOL HIDRATADO,1,R$/13Kg,0.766,0.0,...,1.099,-,0.0,-,-,-,-,-,1,2004
SUL,24,24,2004-05-09,2004-05-15,PARANA,ETANOL HIDRATADO,1,R$/13Kg,0.922,0.0,...,1.199,-,0.0,-,-,-,-,-,1,2004


In [82]:
data_final.groupby('REGIÃO').min()


,Unnamed: 0.1,Unnamed: 0,DATA INICIAL,DATA FINAL,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
REGIÃO,,,,,,,,,,,,,,,,,,,,,
CENTRO OESTE,0,0,2004-05-09,2004-05-15,DISTRITO FEDERAL,ETANOL HIDRATADO,1,R$/13Kg,1.097,0.0,...,1.099,-,0.0,-,-,-,-,-,1,2004
NORDESTE,4,4,2004-05-09,2004-05-15,ALAGOAS,ETANOL HIDRATADO,1,R$/13Kg,0.975,0.0,...,0.999,-,0.0,-,-,-,-,-,1,2004
NORTE,13,13,2004-05-09,2004-05-15,ACRE,ETANOL HIDRATADO,1,R$/13Kg,1.219,0.0,...,1.219,-,0.0,-,-,-,-,-,1,2004
SUDESTE,20,20,2004-05-09,2004-05-15,ESPIRITO SANTO,ETANOL HIDRATADO,1,R$/13Kg,0.766,0.0,...,1.099,-,0.0,-,-,-,-,-,1,2004
SUL,24,24,2004-05-09,2004-05-15,PARANA,ETANOL HIDRATADO,1,R$/13Kg,0.922,0.0,...,1.199,-,0.0,-,-,-,-,-,1,2004


<br/>

Também podemos ter agrupamentos por mais de um atributo.

In [83]:
data_final.head()


,Unnamed: 0.1,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
4,4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,...,1.400,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004


<br/>

`.agg`: **agrega (roda)** uma série de funções para os elementos de um dataframe ou de grupos de um dataframe.

In [84]:
df = pd.DataFrame([[1, 2, 3],
                   [4, 5, 6],
                   [7, 8, 9],
                   [np.nan, np.nan, np.nan]],
                  columns=['A', 'B', 'C'])
df

NameError: name 'np' is not defined

In [85]:
df.agg([sum, min])  # ignora o NaN


NameError: name 'df' is not defined

In [86]:
grupos = data_final.groupby('REGIÃO')
grupos

In [87]:
# Computa o menor e maior valor do 'PREÇO MÉDIO REVENDA' para cada região (grupo)
grupos['PREÇO MÉDIO REVENDA'].agg([min, max])


/tmp/ipython-input-2444597663.py:2: FutureWarning: The provided callable <built-in function min> is currently using SeriesGroupBy.min. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "min" instead.
  grupos['PREÇO MÉDIO REVENDA'].agg([min, max])
/tmp/ipython-input-2444597663.py:2: FutureWarning: The provided callable <built-in function max> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  grupos['PREÇO MÉDIO REVENDA'].agg([min, max])


,min,max
REGIÃO,,
CENTRO OESTE,1.097,99.357
NORDESTE,0.975,79.630
NORTE,1.219,87.270
SUDESTE,0.766,79.192
SUL,0.922,72.088


<h2>6. Ordenação</h2>
<hr/>

In [88]:
notas = pd.DataFrame({
    'nome': ['João', 'Maria', 'José', 'Alice'],
    'idade': [20, 21, 19, 20],
    'nota_final': [5.0, 10.0, 6.0, 10.0]
})
notas

,nome,idade,nota_final
0,João,20,5.0
1,Maria,21,10.0
2,José,19,6.0
3,Alice,20,10.0


`.sort_values()`: ordena valores ao longo de um eixo.

In [89]:
notas.sort_values(by='nota_final')

,nome,idade,nota_final
0,João,20,5.0
2,José,19,6.0
1,Maria,21,10.0
3,Alice,20,10.0


Por padrão, o método retorna uma cópia dos dados ordenados em **ordem crescente (ascendente)**. Podemos alterar isso pelo argumento `ascending`.

In [90]:
notas.sort_values(by='nota_final', ascending=False)

,nome,idade,nota_final
1,Maria,21,10.0
3,Alice,20,10.0
2,José,19,6.0
0,João,20,5.0


<br/>

Podemos ordenar a partir de **mais de uma coluna**:

In [91]:
notas.sort_values(by=['nota_final', 'nome'], ascending=[False, True])


,nome,idade,nota_final
3,Alice,20,10.0
1,Maria,21,10.0
2,José,19,6.0
0,João,20,5.0


Ordena os registros, primeiramente, pela coluna 'nota_final' em **ordem descrente**. <br/>
Então, reordena os registros _"empatados"_, ou seja, com a **mesma nota final**, em _ordem alfabética_ (ordem crescente).

<br/>

Note que o dataframe original **não foi alterado** após a ordenação.

In [92]:
notas

,nome,idade,nota_final
0,João,20,5.0
1,Maria,21,10.0
2,José,19,6.0
3,Alice,20,10.0


Para alterá-lo, use o argumento `inplace=True`:

In [93]:
notas.sort_values(by=['nota_final', 'nome'], ascending=[False, True], inplace=True)
notas

,nome,idade,nota_final
3,Alice,20,10.0
1,Maria,21,10.0
2,José,19,6.0
0,João,20,5.0


<h2>7. Exercícios</h2>
<hr/>

Vamos aplicar os conceitos que vimos em alguns exercícios. <br/>
Para isso, utilizaremos o dataset de _preços de combustíveis no Brasil_.

Como há apenas medições de janeiro a junho para o ano de 2019, resolvemos **remover os dados** deste ano da análise.

In [94]:
data_final

,Unnamed: 0.1,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
4,4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,...,1.400,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
106818,106818,106818,2019-06-23,2019-06-29,NORDESTE,RIO GRANDE DO NORTE,GNV,7,R$/m3,3.574,...,3.690,0.818,0.018,2.756,0,2.756,2.756,0,6,2019
106819,106819,106819,2019-06-23,2019-06-29,SUL,RIO GRANDE DO SUL,GNV,23,R$/m3,3.401,...,3.789,0.95,0.038,2.451,0.402,1.9842,2.8661,0.164,6,2019
106820,106820,106820,2019-06-23,2019-06-29,SUL,SANTA CATARINA,GNV,24,R$/m3,2.912,...,3.499,0.914,0.065,1.998,0,1.9981,1.9981,0,6,2019
106821,106821,106821,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,GNV,52,R$/m3,3.020,...,3.490,0.646,0.076,2.374,0.165,2.0179,2.5093,0.07,6,2019


Note que temos um novo dataframe com 99739 linhas, mas com índices fora desse intervalo. <br/>
Acontece que os registros mantiveram seus índices originais após a query. <br/>

Para resetar os índices de _0 a num_linhas-1_, basta usarmos o método `.reset_index()`.

In [95]:
data_final.query('ANO != 2019')


,Unnamed: 0.1,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
4,4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,...,1.400,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
103064,103064,103064,2018-12-30,2019-01-05,NORDESTE,RIO GRANDE DO NORTE,GNV,10,R$/m3,3.465,...,3.590,0.759,0.017,2.706,0,2.7059,2.7059,0,12,2018
103065,103065,103065,2018-12-30,2019-01-05,SUL,RIO GRANDE DO SUL,GNV,16,R$/m3,3.402,...,3.499,1.296,0.030,2.106,0.249,1.8775,2.3724,0.118,12,2018
103066,103066,103066,2018-12-30,2019-01-05,SUL,SANTA CATARINA,GNV,15,R$/m3,2.944,...,3.199,0.67,0.042,2.274,0.283,1.8738,2.274,0.124,12,2018
103067,103067,103067,2018-12-30,2019-01-05,SUDESTE,SAO PAULO,GNV,41,R$/m3,2.353,...,3.429,0.467,0.192,1.886,0.25,1.64,2.332,0.133,12,2018


In [96]:
df = data_final.query('ANO != 2019')
df

,Unnamed: 0.1,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
4,4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,...,1.400,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
103064,103064,103064,2018-12-30,2019-01-05,NORDESTE,RIO GRANDE DO NORTE,GNV,10,R$/m3,3.465,...,3.590,0.759,0.017,2.706,0,2.7059,2.7059,0,12,2018
103065,103065,103065,2018-12-30,2019-01-05,SUL,RIO GRANDE DO SUL,GNV,16,R$/m3,3.402,...,3.499,1.296,0.030,2.106,0.249,1.8775,2.3724,0.118,12,2018
103066,103066,103066,2018-12-30,2019-01-05,SUL,SANTA CATARINA,GNV,15,R$/m3,2.944,...,3.199,0.67,0.042,2.274,0.283,1.8738,2.274,0.124,12,2018
103067,103067,103067,2018-12-30,2019-01-05,SUDESTE,SAO PAULO,GNV,41,R$/m3,2.353,...,3.429,0.467,0.192,1.886,0.25,1.64,2.332,0.133,12,2018


Os índices agora foram **resetados**. Porém, os índices antigos se transformaram em _uma nova coluna_ chamada 'index'. <br/>
Para removê-la durante o _reset_, basta passarmos o argumento `drop=True`.

In [97]:
grupos = df.groupby('PRODUTO')
grupos



In [98]:
grupos['REGIÃO'].value_counts().to_frame()

count
PRODUTO          REGIÃO             
ETANOL HIDRATADO NORDESTE       6839
                 NORTE          5244
                 CENTRO OESTE   3040
                 SUDESTE        3040
                 SUL            2279
GASOLINA COMUM   NORDESTE       6840
                 NORTE          5319
                 CENTRO OESTE   3040
                 SUDESTE        3040
                 SUL            2280
GLP              NORDESTE       6836
                 NORTE          5319
                 SUDESTE        3039
                 CENTRO OESTE   3037
                 SUL            2280
GNV              NORDESTE       5398
                 SUDESTE        3036
                 SUL            2269
                 CENTRO OESTE   1493
                 NORTE           444
ÓLEO DIESEL      NORDESTE       6840
                 NORTE          5319
                 CENTRO OESTE   3040
                 SUDESTE        3040
                 SUL            2280
ÓLEO DIESEL S10  NORDESTE       2814
                 NORTE          2184
                 SUDESTE        1252
                 CENTRO OESTE   1249
                 SUL             939

### 7.1 Qual a proporção de postos pesquisados para cada combustível em cada região

### 7.2 Como os preços da Gasolina Comum em São Paulo variaram em 2018?

In [99]:
gasolina_sp_2018 = df.query('PRODUTO == "GASOLINA COMUM" and ESTADO == "SAO PAULO" and ANO == 2018')
gasolina_sp_2018.head()

,Unnamed: 0.1,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
95351,95351,95351,2018-01-07,2018-01-13,SUDESTE,SAO PAULO,GASOLINA COMUM,1641,R$/l,3.988,...,4.599,0.411,0.039,3.577,0.104,2.99,3.951,0.029,1,2018
95502,95502,95502,2018-01-14,2018-01-20,SUDESTE,SAO PAULO,GASOLINA COMUM,1648,R$/l,4.002,...,4.499,0.412,0.039,3.59,0.102,2.99,3.9014,0.028,1,2018
95654,95654,95654,2018-01-21,2018-01-27,SUDESTE,SAO PAULO,GASOLINA COMUM,1648,R$/l,4.006,...,4.899,0.41,0.040,3.596,0.102,2.99,3.9014,0.028,1,2018
95806,95806,95806,2018-01-28,2018-02-03,SUDESTE,SAO PAULO,GASOLINA COMUM,1644,R$/l,4.019,...,4.699,0.423,0.039,3.596,0.107,2.924,3.9537,0.03,1,2018
95958,95958,95958,2018-02-04,2018-02-10,SUDESTE,SAO PAULO,GASOLINA COMUM,1646,R$/l,4.004,...,4.597,0.416,0.039,3.588,0.102,2.99,3.9537,0.028,2,2018


In [100]:
gasolina_sp_2018.shape


(52, 22)

#### **Estatísticas Descritivas**

In [101]:
gasolina_sp_2018.describe()


,Unnamed: 0.1,Unnamed: 0,NÚMERO DE POSTOS PESQUISADOS,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,PREÇO MÍNIMO REVENDA,PREÇO MÁXIMO REVENDA,COEF DE VARIAÇÃO REVENDA,MÊS,ANO
count,52.000000,52.000000,52.000000,52.000000,52.000000,52.000000,52.000000,52.000000,52.000000,52.0
mean,99196.096154,99196.096154,1610.211538,4.202769,0.175500,3.671500,4.975750,0.041750,6.615385,2018.0
std,2279.595314,2279.595314,213.627432,0.180672,0.017955,0.163808,0.211198,0.003965,3.470627,0.0
min,95351.000000,95351.000000,102.000000,3.970000,0.150000,3.369000,4.499000,0.034000,1.000000,2018.0
25%,97281.250000,97281.250000,1639.000000,4.008000,0.159750,3.499000,4.899000,0.039000,4.000000,2018.0
50%,99196.500000,99196.500000,1642.000000,4.202000,0.173000,3.679000,4.999000,0.040000,7.000000,2018.0
75%,101116.250000,101116.250000,1644.000000,4.336250,0.189000,3.784000,5.099000,0.044000,9.250000,2018.0
max,103025.000000,103025.000000,1649.000000,4.512000,0.216000,3.969000,5.299000,0.051000,12.000000,2018.0


In [102]:
gasolina_sp_2018['PREÇO MÉDIO REVENDA'].describe().to_frame()


,PREÇO MÉDIO REVENDA
count,52.000000
mean,4.202769
std,0.180672
min,3.970000
25%,4.008000
50%,4.202000
75%,4.336250
max,4.512000


### 7.3 Como os preços da Gasolina Comum e do Etanol em São Paulo variaram em 2018?

In [103]:
df.query('(PRODUTO == "GASOLINA COMUM" or PRODUTO == "ETANOL HIDRATADO") and ESTADO == "SAO PAULO" and ANO == 2018')


,Unnamed: 0.1,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
95270,95270,95270,2018-01-07,2018-01-13,SUDESTE,SAO PAULO,ETANOL HIDRATADO,1664,R$/l,2.833,...,3.399,0.338,0.049,2.495,0.092,2.1842,2.8833,0.037,1,2018
95351,95351,95351,2018-01-07,2018-01-13,SUDESTE,SAO PAULO,GASOLINA COMUM,1641,R$/l,3.988,...,4.599,0.411,0.039,3.577,0.104,2.99,3.951,0.029,1,2018
95421,95421,95421,2018-01-14,2018-01-20,SUDESTE,SAO PAULO,ETANOL HIDRATADO,1661,R$/l,2.868,...,3.399,0.331,0.047,2.537,0.087,2.26,2.8471,0.034,1,2018
95502,95502,95502,2018-01-14,2018-01-20,SUDESTE,SAO PAULO,GASOLINA COMUM,1648,R$/l,4.002,...,4.499,0.412,0.039,3.59,0.102,2.99,3.9014,0.028,1,2018
95573,95573,95573,2018-01-21,2018-01-27,SUDESTE,SAO PAULO,ETANOL HIDRATADO,1662,R$/l,2.874,...,3.399,0.324,0.047,2.55,0.095,1.69,2.9722,0.037,1,2018
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
102724,102724,102724,2018-12-16,2018-12-22,SUDESTE,SAO PAULO,GASOLINA COMUM,1637,R$/l,4.129,...,5.099,0.49,0.049,3.639,0.11,3.35,3.9892,0.03,12,2018
102793,102793,102793,2018-12-23,2018-12-29,SUDESTE,SAO PAULO,ETANOL HIDRATADO,1662,R$/l,2.644,...,3.699,0.305,0.062,2.339,0.094,2.0067,2.7443,0.04,12,2018
102874,102874,102874,2018-12-23,2018-12-29,SUDESTE,SAO PAULO,GASOLINA COMUM,1641,R$/l,4.129,...,4.999,0.484,0.049,3.645,0.107,3.35,3.9892,0.029,12,2018
102944,102944,102944,2018-12-30,2019-01-05,SUDESTE,SAO PAULO,ETANOL HIDRATADO,1661,R$/l,2.643,...,3.699,0.303,0.061,2.34,0.097,1.9072,2.7957,0.041,12,2018


In [104]:
# considerando os preços do etanol e da gasolina juntos, teremos essas estatística descritivas
gasolina_etanol_sp_2018['PREÇO MÉDIO REVENDA'].describe().to_frame()


NameError: name 'gasolina_etanol_sp_2018' is not defined

In [106]:

gasolina_etanol_sp_2018.groupby('PRODUTO')['PREÇO MÉDIO REVENDA'].describe()


NameError: name 'gasolina_etanol_sp_2018' is not defined

<h2>8. Assuntos para continuar os estudos</h2>
<hr/>

- join
- concat
- plot
- data cleaning